bin/tx-prover/src/proxy/mod.rs

```diff
@@ -0,0 +1,193 @@
+        Self::CTX: Send + Sync,
+    {
+        // Check if it's a gRPC request
+        if let Some(content_type) = upstream_request.headers.get("content-type") {
+            if content_type == "application/grpc" {
+                // Ensure the correct host and gRPC headers are forwarded
+                // upstream_request.insert_header("Host", "0.0.0.0").unwrap();
+                upstream_request.insert_header("content-type", "application/grpc").unwrap();
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn request_filter(&self, session: &mut Session, _ctx: &mut Self::CTX) -> Result<bool>
+    where
+        Self::CTX: Send + Sync,
+    {
+        let client_addr = session.client_addr();
+        let user_id = client_addr.map(|addr| addr.to_string());
+
+        // Request ID is a random number
+        let request_id = rand::random::<u64>().to_string();
+        session
+            .req_header_mut()
+            .insert_header("X-Request-ID", request_id)
+            .expect("Failed to insert header");
+
+        // retrieve the current window requests
+        let curr_window_requests = RATE_LIMITER.observe(&user_id, 1);
+
+        if curr_window_requests > MAX_REQ_PER_SEC {
+            // rate limited, return 429
+            let mut header = ResponseHeader::build(429, None).unwrap();
+            header.insert_header("X-Rate-Limit-Limit", MAX_REQ_PER_SEC.to_string()).unwrap();
+            header.insert_header("X-Rate-Limit-Remaining", "0").unwrap();
+            header.insert_header("X-Rate-Limit-Reset", "1").unwrap();
+            session.set_keepalive(None);
+            session.write_response_header(Box::new(header), true).await?;
+            return Ok(true);
+        }
+        Ok(false)
+    }
+
+    async fn logging(&self, session: &mut Session, e: Option<&Error>, _ctx: &mut Self::CTX)
+    where
+        Self::CTX: Send + Sync,
+    {
+        if let Some(e) = e {
+            error!("Error: {:?}", e);
+        }
+
+        // Get the request ID from the session
+        let request_id = session
+            .get_header("X-Request-ID")
+            .expect("Request ID not found")
+            .to_str()
+            .expect("Invalid header value");
+
+        // Remove the completed request from the backend queue
+        // Maybe we can replace this with a read lock and using write only in the moment of the
+        // deletion.
+        let mut ctx_guard = QUEUES.write().await;
+
+        // Get the upstream by checking each backend queue
+        let upstream = ctx_guard
+            .iter()
+            .find(|(_, queue)| queue.contains(&request_id.to_string()))
+            .map(|(upstream, _)| upstream.clone())
+            .expect("Upstream not found");
+
+        // Remove the request ID from the queue for the specific backend
+        if let Some(backend_queue) = ctx_guard.get_mut(&upstream) {
+            if !backend_queue.is_empty() {
+                backend_queue.remove(0);
+            }
+        }
```

Comment:
nit: If this function also removes the request from the queue, maybe I would change its name. `logging` gives me the idea that it's more of a read-only function.

Reply:
I think this is because it's a service trait implementation, so it cannot be renamed.

Reply:
don't love the name too, but this comes from Pingora's `ProxyHttp` trait. Following the [request cycle as seen in the docs](https://github.com/cloudflare/pingora/blob/main/docs/user_guide/phase.md#logging), the `logging` method is called after every request no matter how it went, that's why it was the only place to just do one dequeue. I first tried doing it in the `response_filter` but realized that in any error case it would not dequeue that request and cause basically the workers to stop receiving jobs.
As they mention in the docs, `logging` is the right phase to do post request clean-up.
Though if you have any further concerns we can discuss it